packages/dd-trace/src/collector/span.js

```diff
@@ -0,0 +1,118 @@
+'use strict'
+
+const { channel } = require('dc-polyfill')
+const DatadogSpan = require('../opentracing/span')
+const DatadogCollectorSpanContext = require('./span_context')
+
+const now = performance.now.bind(performance)
+
+const startSegmentChannel = channel('datadog:tracing:segment:start')
+const segmentDiscardChannel = channel('datadog:tracing:segment:discard')
+const startChannel = channel('datadog:tracing:span:start')
+const tagsChannel = channel('datadog:tracing:span:tags')
+const errorChannel = channel('datadog:tracing:span:error')
+const finishChannel = channel('datadog:tracing:span:finish')
+
+let segmentId = 0
+
+class DatadogCollectorSpan extends DatadogSpan {
```

Comment:
This sub-classing is rather difficult to follow. Can we not just put all this logic in-place in `DatadogSpan` and use `channel.hasSubscribers` to skip it? There's a lot of circular dependency between this and the parent class calling methods on this that makes it challenging to understand the flow. I feel like it'd be a lot easier to understand if it was done in the parent class instead, or even with plain functions _called_ from the parent class. 🤔

Reply:
The idea was to isolate the logic to a separate class instead of mixing it with the existing logic. I could instead just put all the code in the same place but it seemed like it would be even _more_ confusing. I felt like keeping them separate also makes it more obvious what is shared versus what isn't.

Reply:
The parent calling a protected method on the child is how most of the tracer works since it comes from an OpenTracing background and OT does that extensively, so I don't think it should be surprising to anyone reading the code.

With that said, I can change it if you really want, but I'm not sure of the best approach then. I don't necessarily want to do it in the same file, because one of the ideas here was to not load this code at all when it's not needed. Not doing this is currently a huge problem in the library especially for serverless, and I didn't want to exacerbate the issue. What approach would you suggest with that in mind?

Reply:
Not going to block on this one, just wanted to point out it's really hard to follow so it would be _ideal_ if we could make it more coherent somehow.

One possibility to enable only loading when you _need_ it is multiple layers of diagnostics channel. You could have the existing class publish events with subscribers that do the transformations these functions are doing and then publish the result to those collector channels. Not sure that would be any clearer, but it _would_ enable optional loading.

We could also consider some sort of dependency injection type thing where we register export strategies on the tracer and the span class could just loop over registered strategies and call each. Then the flow goes in a more singular direction rather than this confusing circular dependency.

Reply:
These sound like great ideas to explore, but in my opinion are out of the scope of this PR. Internals are already loaded in various ways that can end up being hard to follow, and standardizing on a way to do this would be more beneficial than introducing yet another way to load things for this PR specifically.

Reply:
I'm more concerned that piling on more complexity will make it even harder to untangle in the future whenever we get to cleaning that up. The present existence of tech debt doesn't justify making more. 😅